docs/DynamicCasting.md

```diff
@@ -0,0 +1,511 @@
+
+## Existential Metatypes
+
+Protocols can specify constraints and provide default implementations for instances of types.
+They can also specify constraints and provide default implementations for static members of types.
+As described above, casting a regular instance of a type to a protocol type produces a protocol witness instance that exposes only those features required or provided by the protocol.
+Similarly, a type identifier instance (`T.self`) can be cast to a protocol's "existential metatype" to expose just the parts of the type corresponding to the protocol's static requirements.
+
+The existential metatype of a protocol `P` is called `P.Type`.
+(Recall that for a non-protocol type `T`, the expression `T.Type` refers to the regular metatype.
+Non-protocol types do not have existential metatypes.
+For a generic variable `G`, the expression also refers to the regular metatype, even if the generic variable is bound to a protocol.
+There is no mechanism in Swift to refer to the existential metatype via a generic variable.)
+
+Example
+```
+protocol P {
+   var ivar: Int { get }
+   static svar: Int { get }
+}
+struct S: P {
+   let ivar = 1
+   static let svar = 2
+}
+S().ivar // 1
+S.self.svar // 2
+(S() as! P).ivar // 1
+(S.self as! P.Type).svar // 2
+```
+
+Invariants
+* If `T` conforms to `P` and `t` is an instance of `T`, then `t is P`, and `T.self is P.Type`
+* Since every type `T` conforms to `Any`, `T.self is Any.Type` is always true
+* `Any` self-conforms:  `Any.self is Any.Type == true`
+
+### Note: "Self conforming" protocols
+
+As mentioned above, a protocol definition for `P` implicitly defines associated types `P.Type` (the existential metatype) and `P.Protocol` (the metatype).
+It also defines an associated type called `P` which is the type of a container that can hold any object whose concrete type conforms to the protocol `P`.
+
+A protocol is "self conforming" if the container type `P` conforms to the protocol `P`.
+This is equivalent to saying that `P.self` is an instance of `P.Type`.
+(Remember that it is always an instance of `P.Protocol`.)
+
+This is a concern for Swift because of the following construct, which attempts to invoke a generic `f` in a situation where the concrete instance clearly conforms to `P` but is represented as a `P` existential:
+```
+func f<T:P>(t: T) { .. use t .. }
+let a : P = something
+f(a)
+```
+This construct is valid only if `T` conforms to `P` when `T = P`; that is, if `P` self-conforms.
+
+Note that any protocol that specifies static methods, static properties, associated types, or initializers cannot possibly be self-conforming.
+As of Swift 5.2, there are in fact only two kinds of self-conforming protocols:
+* `Any` must be self-conforming since every `T.self` is an instance of `Any.Type`
+* Objective-C protocols that have no static requirements are considered self-conforming
```

Comment:
`Error` [self-conforms](https://github.com/apple/swift/pull/20629) as well.

Reply:
Are you sure?  I don't see that behavior in Playground here.

Reply:
Yeah (see the linked PR), it was added in Swift 5 to support the new `Result` type, although I’m not sure if it behaves differently in some situations compared to `Any`/`@objc` as mentioned above.

Reply:
I've clarified my expectations for self-conforming protocols.  Specifically,
* `P.self is P.Type == true`
* `func f<T: P>(...)` can be instantiated for `T = P`
* a generic type `G<T:P>` can be instantiated for `T= P`
It looks like `Error` is indeed intended to be self-conforming, so I've added it to the list of such protocols.